Triplet Attention

iyaja opened the pull request:
This adds support for the triplet attention module as in https://github.com/LandskapeAI/triplet-attention.

@digantamisra98

rwightman commented:
@iyaja @digantamisra98 thanks for the PR, I'll give it a try after I finish merging a fairly significant branch (may cause some conflicts that need to be fixed here).

One thing I'd like changed though is a reduction in the number of models, way too many untrained models for a new model variant that is as of yet untested/untrained by me. Each model def is CI tested and throwing so many in there just blows up the test time with limited utility when there are no weights, it'd take a LONG time to produce weights for all of those.

 I'd like to reduce the added model definition count to maybe 3-4 models, and no 'huge' ones yet. I'd definitely like to see a `50D` variant in there as that's one of the more common baselines I have comparisons for with recent training hparams. For the others at least one 'BasicBlock` model like 34, possibly one larger/higher capacity like a 101. I'd make them all D variants.

```
triplet_resnet34d
triplet_resnet50d
triplet_resnet101d
```

BTW, for the naming, you don't strictly have to follow the 'Module` postfix for the module. The existing attentions all have Module at the end because they were named based on a short acryonym like SE, ECA, etc that makes for a poor class name. `TripletAttention` would be fine. If you did want a shorter name for the models,etc you could use `ta`

`ta_resnet50d` + `TaModule` or `ta_resnet50d` + `TripletAttn` ... any of the combos work really, just throwing out some ideas.

And, when ready please squash commits.

Those are my few cents. Thanks!

iyaja commented:
@rwightman thanks for your feedback!

I renamed the module itself to `TripletAttention` and the models are still `triplet_resnetXX`. Do you think that's ok? It's similar to what we have on our [official repository](https://github.com/LandskapeAI/triplet-attention).

I used the D variants as you suggested (please let me know if I've implemented that incorrectly), and removed everything but the following models:

```
triplet_resnet18d
triplet_resnet50d
triplet_resnet101d
```

In our paper, we have results/benchmarks for ResNet18, ResNet50, and ResNet101 primarily, so I used `triplet_resnet18d` instead of `triplet_resnet34d`.

Also, there was another small change I had to make to `TripletAttention`'s `forward` function to pass the torchscript test.

rwightman commented:
@iyaja @digantamisra98 thanks for the updates... I found some time to try out the models this weekend. The runtime characteristics are problematic. The model is less than half the throughput of ECA or SE ResNets of same size for inference and slower than 1/3 the throughput in training. In terms of trying to find equivalent for training, I can train an SEResNet152-D or ResNet200D 20% faster than the triplet_resnet50d. Those models can hit 16% top-1 error.

Any way to improve runtime throughput and memory usage (uses about 2x the GPU ram for training).

digantamisra98 commented:
@rwightman Apologies for the delay in response. The way Triplet is structured is unfortunately causing redundant full tensor size level operations (permutation). We obviously are collectively looking at ways to optimize it further however as of now I'm afraid Triplet will cause the drop in throughput which you experienced.

sidml commented:
@digantamisra98 
Any updates ?

digantamisra98 commented:
@sidml On pure python level, we haven't yet come up with a way that we can tone down on the performance bottleneck incurred by Triplet Attention. However, it is possible to do so by writing custom CUDA kernels but since my familiarity with it is rather limited, I can't say with definitive confidence on when and if this will be resolved. Hopefully someone expert in low level kernel writing can take this up.

rwightman commented:
@digantamisra98 @iyaja thanks for the PR & efforts but I think it's time to close this. The performance impact over other options is significant, given that, I don't see myself retraining with better hparams to make them competitive with other models. Custom CUDA might be an answer, but I've been avoiding maintaining such code here for maintenance / overhead reasons.
